src/cockpit/polkit.py

```diff
@@ -0,0 +1,154 @@
+                     action_id, message, icon_name, details, cookie, identities)
+        # only support authentication as ourselves, as we don't yet have the
+        # protocol plumbing and UI to select an admin user
+        my_uid = os.geteuid()
+        for (auth_type, subject) in identities:
+            if auth_type == 'unix-user' and 'uid' in subject and subject['uid'].value == my_uid:
+                logger.debug('Authentication subject %s matches our uid %d', subject, my_uid)
+                break
+        else:
+            logger.warning('Not supporting authentication as any of %s', identities)
+            return
+
+        user_name = pwd.getpwuid(my_uid).pw_name
+        process = await asyncio.create_subprocess_exec(HELPER_PATH, user_name, cookie,
+                                                       stdin=asyncio.subprocess.PIPE,
+                                                       stdout=asyncio.subprocess.PIPE)
+        try:
+            await self._communicate(process)
+        except asyncio.CancelledError:
+            logger.debug('Cancelled authentication')
+            process.terminate()
+        finally:
+            res = await process.wait()
+            logger.debug('helper exited with code %i', res)
+
+    async def _communicate(self, process: asyncio.subprocess.Process) -> None:
+        assert process.stdin
+        assert process.stdout
+
+        messages: List[str] = []
+
+        async for line in process.stdout:
+            logger.debug('Read line from helper: %s', line)
+            command, _, value = line.strip().decode().partition(' ')
+
+            # usually: PAM_PROMPT_ECHO_OFF Password: \n
+            if command.startswith('PAM_PROMPT'):
+                # Don't pass this to the UI if it's "Password" (the usual case),
+                # so that superuser.py uses the translated default
+                if value.startswith('Password'):
+                    value = ''
+
+                # flush out accumulated info/error messages
+                passwd = await self.responder.do_askpass('\n'.join(messages), value, '')
+                messages.clear()
+                if passwd is None:
+                    logger.debug('got PAM_PROMPT %s, but do_askpass returned None', value)
+                    raise asyncio.CancelledError('no password given')
+                logger.debug('got PAM_PROMPT %s, do_askpass returned a password', value)
+                process.stdin.write(passwd.encode())
+                process.stdin.write(b'\n')
+                del passwd  # don't keep this around longer than necessary
+                await process.stdin.drain()
+                logger.debug('got PAM_PROMPT, wrote password to helper')
+            elif command in ('PAM_TEXT_INFO', 'PAM_ERROR'):
+                messages.append(value)
+            elif command == 'SUCCESS':
+                logger.info('Authentication succeeded')
+                break
+            elif command == 'FAILURE':
```

Comment:
Do we get a message along with this one, or should we raise the previously collected messages as our error message?  I guess that would (eventually) be a `PeerError` of some kind...

Reply:
This is just a single value, like `SUCCESS`. Unfortunately there is very little dialog here. For good password:
```
PAM_PROMPT_ECHO_OFF Password: \n
SUCCESS
```
and for wrong password:
```
PAM_PROMPT_ECHO_OFF Password: \n
FAILURE
```
The PAM helper logs `polkit-agent-helper-1: pam_authenticate failed: Authentication failure` (but that's outside of that dialog), and then exits with code 1. The only remaining message is pkexec itself writing "Not authorized This incident has been reported", which we then show in the UI:

![image](https://github.com/cockpit-project/cockpit/assets/200109/92db5adb-9b09-4815-830c-da1786b080d4)

That's why I didn't bother much about the messages, as they are rather hard to test.

I tried to raise a PeerError there, but it just crashes the bridge as nothing catches it:

```
asyncio-ERROR: Exception in callback BusMessage._coroutine_task_complete(<cockpit._ven...x7f138b9803c8>)(<Task finishe...ion-failed',)>)
handle: <Handle BusMessage._coroutine_task_complete(<cockpit._ven...x7f138b9803c8>)(<Task finishe...ion-failed',)>)>
Traceback (most recent call last):
  File "/usr/lib64/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "/usr/lib/python3.6/site-packages/cockpit/_vendor/systemd_ctypes/bus.py", line 179, in _coroutine_task_complete
    self.reply_method_function_return_value(out_type, task.result())
  File "/usr/lib/python3.6/site-packages/cockpit/polkit.py", line 57, in begin_authentication
    await self._communicate(process)
  File "/usr/lib/python3.6/site-packages/cockpit/polkit.py", line 101, in _communicate
    raise PeerError('authentication-failed')
cockpit.peer.PeerError: authentication-failed
```

Reply:
Could we capture the pam messages via stderr?

Reply:
We capture stderr of pkexec, which is what generates that "Not authorized .. reported" message. I don't think we want to  capture the journal here.